new xdot.py package and necessary updates to GUI libs

lee218llnl opened the pull request.

lee218llnl commented:
Can someone take a look at this?

scheibelp commented:
Yes - I'll have a look today. Thanks for the ping!

adamjstewart commented on var/spack/repos/builtin/packages/gtkplus/package.py:
```diff
@@ -30,7 +30,9 @@ class Gtkplus(AutotoolsPackage):
-
+    version('3.20.10',
+            'e81da1af1c5c1fee87ba439770e17272fa5c06e64572939814da406859e56b70',
+            url='http://ftp.gnome.org/pub/gnome/sources/gtk+/3.20/gtk+-3.20.10.tar.xz')
```
Can you override `url_for_version` so that this version-specific URL becomes unnecessary?

adamjstewart commented on var/spack/repos/builtin/packages/py-pycairo/package.py:
```diff
@@ -0,0 +1,48 @@
+import os
+
+
+class PyPycairo(PythonPackage):
```
There is already a `py-py2cairo` package that appears to be a duplicate. I agree that the name should be `py-pycairo`. I would prefer to download directly from PyPI.

lee218llnl replied:
ugh, the py-py2cairo package name is a bit misleading. Perhaps we can make a separate PR to rename it and merge to py-pycairo and then modify all the packages that depend on it? I fear there will be some major compatibility issues that will arise since py2cairo is for Python 2 and pycairo is for Python 3.

adamjstewart commented on var/spack/repos/builtin/packages/py-pycairo/package.py:
```diff
@@ -0,0 +1,48 @@
+
+    extends('python')
+
+    depends_on('python', type=('build', 'run'))
```
Both of these lines are unnecessary, they are already included in `PythonPackage`

adamjstewart commented on var/spack/repos/builtin/packages/py-pygobject/package.py:
```diff
@@ -25,30 +25,48 @@
     url      = "http://ftp.gnome.org/pub/GNOME/sources/pygobject/2.28/pygobject-2.28.6.tar.bz2"
 
+    version('3.28.3', 'c70b2d8c492853fca5d7fd556ce39d36',
+            url='https://files.pythonhosted.org/packages/e0/e8/1e4f21800015a9ca153969e85fc29f7962f8f82fc5dbc1ecbdeb9dc54c75/PyGObject-3.28.3.tar.gz')
```
Are there no single URLs where every version can be found?

adamjstewart commented on var/spack/repos/builtin/packages/py-pygobject/package.py:
```diff
@@ -25,30 +25,48 @@
     depends_on("libffi")
     depends_on('glib')
-    depends_on('py-py2cairo', type=('build', 'run'))
+    depends_on('python@2:2.99', when='@2:2.99')
```
`type=('build', 'run')`?

adamjstewart commented on var/spack/repos/builtin/packages/py-pygobject/package.py:
```diff
@@ -25,30 +25,48 @@
-    depends_on('py-py2cairo', type=('build', 'run'))
+    depends_on('python@2:2.99', when='@2:2.99')
+    depends_on('py-pycairo', type=('build', 'run'), when='@3:')
+    depends_on('py-py2cairo', type=('build', 'run'), when='@2:2.99')
```
Are these actually different packages?

adamjstewart commented on var/spack/repos/builtin/packages/py-pygobject/package.py:
```diff
@@ -25,30 +25,48 @@
     # for https://bugzilla.gnome.org/show_bug.cgi?id=668522
     patch('pygobject-2.28.6-gio-types-2.32.patch', when='@2.28.6')
 
+    # pygobject version 2 requires an autotools build
```
If we're dealing with multiple build systems it might be better to subclass from `Package`.

lee218llnl replied:
@adamjstewart the older version (2.28.*) are autotools packages and my guess is that moving forward this will be a python package. My thought was that it would be better to make the spack package a pythonpackage for when/if we deprecate the older versions. This would save adding all of the python build logic that is built in to the pythonpackage. If you feel strongly about making it subclass Package, then I can do that, though.

adamjstewart commented on var/spack/repos/builtin/packages/py-pygtk/package.py:
```diff
@@ -37,6 +37,11 @@ class PyPygtk(AutotoolsPackage):
+    depends_on('atk@:2.20.0')
+    # PyGTK requires python 2
+    # Use py-pygobject@3: for GTK bindings for python 3
+    depends_on('python@2:2.99')
```
`type=('build', 'run')`?

adamjstewart commented on var/spack/repos/builtin/packages/py-xdot/package.py:
```diff
@@ -0,0 +1,83 @@
+
+    import_modules = ['xdot']
+
+    extends('python')
```
Not necessary, already present in `PythonPackage`.

adamjstewart commented on var/spack/repos/builtin/packages/py-xdot/package.py:
```diff
@@ -0,0 +1,83 @@
+    version('0.9.1', commit="0fa629166989576b05d509c7ef0329c0f7655190")
+    version('0.9', '19c78311d73b0f9ea059a6febf42eeea')
+
+    depends_on('py-setuptools', type=('build', 'run'))
```
Is this actually needed at run-time?

adamjstewart commented on var/spack/repos/builtin/packages/py-xdot/package.py:
```diff
@@ -0,0 +1,83 @@
+    version('0.9', '19c78311d73b0f9ea059a6febf42eeea')
+
+    depends_on('py-setuptools', type=('build', 'run'))
+    depends_on('python@3:')
```
`type=('build', 'run')`?

adamjstewart commented on var/spack/repos/builtin/packages/py-xdot/package.py:
```diff
@@ -0,0 +1,83 @@
+            join_path(spec['pango'].prefix.lib, 'girepository-1.0'),
+            join_path(spec['atk'].prefix.lib64, 'girepository-1.0'),
+            join_path(spec['gdk-pixbuf'].prefix.lib, 'girepository-1.0'),
+            join_path(spec['gtkplus'].prefix.lib, 'girepository-1.0'))
```
These should probably be accessed via `spec['...'].libs` as `lib` vs `lib64` may depend on OS and whether it was externally installed or not.

lee218llnl replied:
@adamjstewart  all 4 of those packages only produce versioned libraries (i.e., libgtk-3.so and libpango-1.0.so, etc.). Any way to work around this or should I leave the prefix.lib[64] specifications?

alalazo replied:
Can we also postpone this to a following PR?

lee218llnl replied:
The join_path calls are specific to the xdot package, so I'm hoping we can keep this in this PR. I'm not sure whether other packages would need to do something similar.

Regarding prefix.lib and prefix.lib64, yes, I propose that a separate issue be raised to do a better job finding versioned .so files.

adamjstewart commented on var/spack/repos/builtin/packages/py-xdot/package.py:
```diff
@@ -0,0 +1,83 @@
+                    "import sys\nimport os\nos.environ['GI_TYPELIB_PATH']" +
+                    " = '%s'" % repo_paths, dst)
+        # regenerate the byte-compiled __init__.py
+        python3 = which('python3')
```
`spec['python'].command`

lee218llnl commented:
@adamjstewart I addressed all of your URL, Python package, and build type issues in the latest (squashed) commit. I did not address the pycairo vs py2cairo issue, which I am hoping can be done as a separate PR. I also did not fix the ".libs" vs ".prefix.lib[64]" specification, as the .libs routine does not find the directory for the packages in question. Please let me know how this now looks to you.

lee218llnl commented:
I just pushed some more modifications. py-xdot requires py-setuptools at run time to avoid an import error for pkg_resources. I also modified at-spi2-core with some additional dependences that were needed for ppc64le. Finally, I rebased on develop to pick up Python 3.7.0 package fixes w.r.t. libffi.

lee218llnl commented:
Can someone take another look at this and merge it if it looks OK?

lee218llnl commented:
Attempt 2: Can someone take another look at this and merge it if it looks OK?

alalazo commented:
@lee218llnl I'll have a look and eventually merge this by tomorrow.

alalazo commented on var/spack/repos/builtin/packages/py-pycairo/package.py:
```diff
@@ -0,0 +1,46 @@
+    version('1.17.1', '34c1ee106655b450c4bd57e29371a4a7')
+
+    depends_on('cairo@1.2.0:')
+    depends_on('pkgconfig', type='build')
```
Looking at the package I think this at least:
```python
depends_on('py-setuptools', type='build')
```
too. Not sure if it is also a `run` dependency.

lee218llnl replied:
Added py-setuptools as a build dependency in the latest squashed commit.

lee218llnl commented on var/spack/repos/builtin/packages/mesa/package.py:
```diff
@@ -205,3 +205,8 @@ def libs(self):
             if libs:
                 return libs
+
+    @when('^python@3:')
```
Yes, it looks like it is only an issue with python 3

lee218llnl commented:
@alalazo, I think I addressed all of your comments. There will be a few pending issues that can be addressed outside of this PR. Once the Travis checks are done can you merge?
